File: main-workshop/services/dotnet/ride-completion-service/Services/RideCompletionService.cs
using Amazon.DynamoDBv2;
using Amazon.DynamoDBv2.Model;
using System.Text.Json;

[assembly: LambdaSerializer(typeof(Amazon.Lambda.Serialization.SystemTextJson.DefaultLambdaJsonSerializer))]


public class RideCompletionService
{
    private readonly IAmazonDynamoDB _dynamoDb;
    private readonly string _driversTableName;
    private readonly string _ridesTableName;

    public RideCompletionService()
    {
        _dynamoDb = new AmazonDynamoDBClient();
        _driversTableName = Environment.GetEnvironmentVariable("DRIVERS_TABLE_NAME") ?? "drivers";
        _ridesTableName = Environment.GetEnvironmentVariable("RIDES_TABLE_NAME") ?? "rides";
    }

    public async Task<RideCompletionResult> ProcessPaymentCompletedEventAsync(
        CloudWatchEvent<PaymentCompletedEvent> cloudWatchEvent)
    {
        var result = new RideCompletionResult();

        try
        {
            // Determine ride status based on event type
            var isPaymentFailed = cloudWatchEvent.DetailType == "PaymentFailed";
            var rideStatus = isPaymentFailed ? "payment_failed" : "completed";

            // Get event detail
            var paymentCompletedEvent = cloudWatchEvent.Detail;
            if (paymentCompletedEvent == null)
            {
                throw new Exception("Failed to deserialize payment event");
            }

            // Validate required fields
            if (string.IsNullOrEmpty(paymentCompletedEvent.RideId) ||
                string.IsNullOrEmpty(paymentCompletedEvent.DriverId) ||
                string.IsNullOrEmpty(paymentCompletedEvent.PaymentId))
            {
                throw new ArgumentException("Required fields are missing from event");
            }

            result.PaymentId = paymentCompletedEvent.PaymentId;
            result.RideId = paymentCompletedEvent.RideId;
            result.RiderId = paymentCompletedEvent.RiderId;
            result.DriverId = paymentCompletedEvent.DriverId;
            result.PaymentMethod = paymentCompletedEvent.PaymentMethod;
            result.Amount = paymentCompletedEvent.Amount;

            // Update ride status
            try
            {
                await UpdateRideStatusAsync(paymentCompletedEvent.RideId, rideStatus);
                result.RideUpdateSuccessful = true;
            }
            catch (InvalidOperationException ex) when (ex.Message.Contains("not found"))
            {
                // Ride doesn't exist (e.g., test/synthetic data) - log and continue
                result.RideUpdateSuccessful = false;
                result.ErrorType = "RideNotFound";
                result.ErrorMessage = ex.Message;
                // Don't throw - this is expected for test data
            }
            catch (Exception ex)
            {
                result.RideUpdateSuccessful = false;
                result.ErrorType = "RideUpdateFailed";
                result.ErrorMessage = ex.Message;
                throw;
            }

            // Update driver status to available
            try
            {
                await UpdateDriverStatusAsync(paymentCompletedEvent.DriverId, DriverStatus.Available);
                result.DriverUpdateSuccessful = true;
            }
            catch (Exception ex)
            {
                result.DriverUpdateSuccessful = false;
                result.ErrorType = "DriverUpdateFailed";
                result.ErrorMessage = ex.Message;

                // Don't throw if ride update was successful - partial success
                if (!result.RideUpdateSuccessful)
                {
                    throw;
                }
            }

            // Consider success if driver update succeeded, even if ride doesn't exist (test data)
            result.Success = result.DriverUpdateSuccessful && 
                           (result.RideUpdateSuccessful || result.ErrorType == "RideNotFound");
            return result;
        }
        catch (Exception ex)
        {
            result.Success = false;
            if (string.IsNullOrEmpty(result.ErrorType))
            {
                result.ErrorType = "UnexpectedError";
                result.ErrorMessage = ex.Message;
            }

            throw;
        }
    }

    private async Task UpdateRideStatusAsync(string rideId, string status)
    {
        if (string.IsNullOrEmpty(rideId))
        {
            throw new ArgumentException("RideId cannot be null or empty", nameof(rideId));
        }

        if (string.IsNullOrEmpty(status))
        {
            throw new ArgumentException("Status cannot be null or empty", nameof(status));
        }

        try
        {
            var updateRequest = new UpdateItemRequest
            {
                TableName = _ridesTableName,
                Key = new Dictionary<string, AttributeValue>
                {
                    ["rideId"] = new AttributeValue { S = rideId }
                },
                UpdateExpression = "SET #status = :status, #updatedAt = :updatedAt",
                ExpressionAttributeNames = new Dictionary<string, string>
                {
                    ["#status"] = "status",
                    ["#updatedAt"] = "updatedAt"
                },
                ExpressionAttributeValues = new Dictionary<string, AttributeValue>
                {
                    [":status"] = new AttributeValue { S = status },
                    [":updatedAt"] = new AttributeValue { S = DateTime.UtcNow.ToString("O") }
                },
                ConditionExpression = "attribute_exists(rideId)",
                ReturnValues = ReturnValue.UPDATED_NEW
            };

            await _dynamoDb.UpdateItemAsync(updateRequest);
        }
        catch (ConditionalCheckFailedException ex)
        {
            throw new InvalidOperationException($"Ride with ID {rideId} not found", ex);
        }
        catch (Exception ex)
        {
            throw new InvalidOperationException($"Failed to update ride status for ride {rideId}: {ex.Message}", ex);
        }
    }

    private async Task UpdateDriverStatusAsync(string driverId, string status)
    {
        if (string.IsNullOrEmpty(driverId))
        {
            throw new ArgumentException("DriverId cannot be null or empty", nameof(driverId));
        }

        if (string.IsNullOrEmpty(status))
        {
            throw new ArgumentException("Status cannot be null or empty", nameof(status));
        }

        try
        {
            var updateRequest = new UpdateItemRequest
            {
                TableName = _driversTableName,
                Key = new Dictionary<string, AttributeValue>
                {
                    ["driverId"] = new AttributeValue { S = driverId }
                },
                UpdateExpression = "SET #status = :status, #updatedAt = :updatedAt",
                ExpressionAttributeNames = new Dictionary<string, string>
                {
                    ["#status"] = "status",
                    ["#updatedAt"] = "updatedAt"
                },
                ExpressionAttributeValues = new Dictionary<string, AttributeValue>
                {
                    [":status"] = new AttributeValue { S = status },
                    [":updatedAt"] = new AttributeValue { S = DateTime.UtcNow.ToString("O") }
                },
                ConditionExpression = "attribute_exists(driverId)",
                ReturnValues = ReturnValue.UPDATED_NEW
            };

            await _dynamoDb.UpdateItemAsync(updateRequest);
        }
        catch (ConditionalCheckFailedException ex)
        {
            throw new InvalidOperationException($"Driver with ID {driverId} not found", ex);
        }
        catch (Exception ex)
        {
            throw new InvalidOperationException($"Failed to update driver status for driver {driverId}: {ex.Message}",
                ex);
        }
    }
}

File: main workshop/services/typescript/ride-completion-service/src/services/RideCompletionService.ts
import { DynamoDBClient } from '@aws-sdk/client-dynamodb';
import { DynamoDBDocumentClient, UpdateCommand } from '@aws-sdk/lib-dynamodb';
import { captureAWSv3Client } from 'aws-xray-sdk-core';
import type { PaymentCompletedEvent, RideCompletionResult } from '../models';

export class RideCompletionService {
  private readonly dynamoDb: DynamoDBDocumentClient;
  private readonly driversTableName: string;
  private readonly ridesTableName: string;

  constructor() {
    const client = new DynamoDBClient({});
    this.dynamoDb = captureAWSv3Client(
      DynamoDBDocumentClient.from(client)
    );
    this.driversTableName = process.env.DRIVERS_TABLE_NAME || 'drivers';
    this.ridesTableName = process.env.RIDES_TABLE_NAME || 'rides';
  }

  async processPaymentCompletedEvent(
    event: PaymentCompletedEvent,
    detailType: string
  ): Promise<RideCompletionResult> {
    const result: RideCompletionResult = {
      paymentId: '',
      rideId: '',
      riderId: '',
      driverId: '',
      paymentMethod: '',
      amount: 0,
      rideUpdateSuccessful: false,
      driverUpdateSuccessful: false,
      success: false,
      errorMessage: '',
      errorType: '',
    };

    try {
      // Determine ride status based on event type
      const isPaymentFailed = detailType === 'PaymentFailed';
      const rideStatus = isPaymentFailed ? 'payment_failed' : 'completed';

      // Validate required fields
      if (!event.rideId || !event.driverId || !event.paymentId) {
        throw new Error('Required fields are missing from event');
      }

      result.paymentId = event.paymentId;
      result.rideId = event.rideId;
      result.riderId = event.riderId;
      result.driverId = event.driverId;
      result.paymentMethod = event.paymentMethod;
      result.amount = event.amount;

      // Update ride status
      try {
        await this.updateRideStatus(event.rideId, rideStatus);
        result.rideUpdateSuccessful = true;
      } catch (error) {
        result.rideUpdateSuccessful = false;
        const errorMessage = error instanceof Error ? error.message : 'Unknown error';
        
        // Check if ride doesn't exist (test/synthetic data)
        if (errorMessage.includes('not found')) {
          result.errorType = 'RideNotFound';
          result.errorMessage = errorMessage;
          // Don't throw - this is expected for test data
        } else {
          result.errorType = 'RideUpdateFailed';
          result.errorMessage = errorMessage;
          throw error;
        }
      }

      // Update driver status to available
      try {
        await this.updateDriverStatus(event.driverId, 'available');
        result.driverUpdateSuccessful = true;
      } catch (error) {
        result.driverUpdateSuccessful = false;
        result.errorType = 'DriverUpdateFailed';
        result.errorMessage =
          error instanceof Error ? error.message : 'Unknown error';

        // Don't throw if ride update was successful - partial success
        if (!result.rideUpdateSuccessful) {
          throw error;
        }
      }

      // Consider success if driver update succeeded, even if ride doesn't exist (test data)
      result.success =
        result.driverUpdateSuccessful && 
        (result.rideUpdateSuccessful || result.errorType === 'RideNotFound');
      return result;
    } catch (error) {
      result.success = false;
      if (!result.errorType) {
        result.errorType = 'UnexpectedError';
        result.errorMessage =
          error instanceof Error ? error.message : 'Unknown error';
      }
      throw error;
    }
  }

  private async updateRideStatus(
    rideId: string,
    status: string
  ): Promise<void> {
    if (!rideId) {
      throw new Error('RideId cannot be null or empty');
    }

    if (!status) {
      throw new Error('Status cannot be null or empty');
    }

    try {
      const command = new UpdateCommand({
        TableName: this.ridesTableName,
        Key: {
          rideId: rideId,
        },
        UpdateExpression: 'SET #status = :status, #updatedAt = :updatedAt',
        ExpressionAttributeNames: {
          '#status': 'status',
          '#updatedAt': 'updatedAt',
        },
        ExpressionAttributeValues: {
          ':status': status,
          ':updatedAt': new Date().toISOString(),
        },
        ConditionExpression: 'attribute_exists(rideId)',
        ReturnValues: 'UPDATED_NEW',
      });

      await this.dynamoDb.send(command);
    } catch (error: any) {
      if (error.name === 'ConditionalCheckFailedException') {
        throw new Error(`Ride with ID ${rideId} not found`);
      }
      throw new Error(
        `Failed to update ride status for ride ${rideId}: ${error.message}`
      );
    }
  }

  private async updateDriverStatus(
    driverId: string,
    status: string
  ): Promise<void> {
    if (!driverId) {
      throw new Error('DriverId cannot be null or empty');
    }

    if (!status) {
      throw new Error('Status cannot be null or empty');
    }

    try {
      const command = new UpdateCommand({
        TableName: this.driversTableName,
        Key: {
          driverId: driverId,
        },
        UpdateExpression: 'SET #status = :status, #updatedAt = :updatedAt',
        ExpressionAttributeNames: {
          '#status': 'status',
          '#updatedAt': 'updatedAt',
        },
        ExpressionAttributeValues: {
          ':status': status,
          ':updatedAt': new Date().toISOString(),
        },
        ConditionExpression: 'attribute_exists(driverId)',
        ReturnValues: 'UPDATED_NEW',
      });

      await this.dynamoDb.send(command);
    } catch (error: any) {
      if (error.name === 'ConditionalCheckFailedException') {
        throw new Error(`Driver with ID ${driverId} not found`);
      }
      throw new Error(
        `Failed to update driver status for driver ${driverId}: ${error.message}`
      );
    }
  }
}
